### PR TITLE
MVC Controllers, Views & Models

### PR DESCRIPTION
**Pull request by j1nma**
## Summary
* I followed the MVC Controller card to learn about its behaviour and how routing is managed. The tutorial's HelloWorldController was implemented with 2 simple endpoints.

## Trello Card Link
https://trello.com/c/TDxQ5NpO/2-mvc-controllers

## Screenshots
image

image

### CODE REVIEW

**Comment by j1nma** [comment]
## Summary
* I followed the MVC Views. Changed views and layout pages. Changed the title and menu link in the layout file. Passed data from the Controller to the View.

## Trello Card Link
https://trello.com/c/knsFNSfT/3-mvc-views

## Screenshots
image

image

image

**Comment by j1nma** [comment]
## Summary
* I followed the MVC Models card. Made DB configuration as in [here](https://gist.github.com/gzamudio/424f50d7ff3f1df6c12260b851f722b3) and [here](https://gist.github.com/gzamudio/4aee3f1142d19b1eeb9ffa71442e1bb4). Generated Movie model.

## Trello Card Link
https://trello.com/c/pgML8LgN/4-mvc-models

## Screenshots
image

File: src/Controllers/HelloWorldController.cs [hunk 1]
```diff
@@ -1,23 +1,17 @@
...
+        public IActionResult Index () {
+            return View ();
         }
 
-        // GET: /HelloWorld/Welcome/ 
-        // Requires using System.Text.Encodings.Web;
-        public string Welcome(string name, int ID = 1)
-        {
-            return HtmlEncoder.Default.Encode($"Hello {name}, ID: {ID}");
+        public IActionResult Welcome (string name, int numTimes = 1) {
```
Comment (gzamudio) [inline src/Controllers/HelloWorldController.cs]: we use brackets in next line

File: src/Controllers/HelloWorldController.cs [hunk 2]
```diff
@@ -1,23 +1,17 @@
...
+            ViewData["Message"] = "Hello " + name;
+            ViewData["NumTimes"] = numTimes;
+
+            return View ();
         }
     }
 }
```
Comment (gzamudio) [inline src/Controllers/HelloWorldController.cs]: add empty line

File: src/Controllers/HomeController.cs [hunk 3]
```diff
@@ -1,43 +1,45 @@
...
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.AspNetCore.Mvc.Localization;
+using testing_net.Models;
+
+namespace testing_net.Controllers {
```
Comment (gzamudio) [inline src/Controllers/HomeController.cs]: by convention, we use brackets on next line, not in namespace or class or method declaration

File: src/Controllers/MoviesController.cs [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+namespace testing_net.Controllers
+{
+    public class MoviesController : Controller
+    {
+        public IActionResult Index()
+        {
+            return View();
+        }
+    }
+}
```
Comment (gzamudio) [inline src/Controllers/MoviesController.cs]: add empty line

File: src/Resources/en-US.json [hunk 5]
```diff
@@ -0,0 +1,5 @@
+{
+    "DescriptionPage": "Your application description page.",
+    "ContactPage": "Your contact page.",
+    "HomePage": "Home Page"
+}
```
Comment (gzamudio) [inline src/Resources/en-US.json]: add empty line

File: src/Resources/es-MX.json [hunk 6]
```diff
@@ -0,0 +1,5 @@
+{
+    "DescriptionPage": "Página de descripción de su aplicación",
+    "ContactPage": "Página de contacto",
+    "HomePage": "Página de inicio"
+}
```
Comment (gzamudio) [inline src/Resources/es-MX.json]: add empty line

File: src/Resources/fr-FR.json [hunk 7]
```diff
@@ -0,0 +1,5 @@
+{
+    "DescriptionPage": "Page de description d'application",
+    "ContactPage": "Page de contact",
+    "HomePage": "Page d'accueil"
+}
```
Comment (gzamudio) [inline src/Resources/fr-FR.json]: add empty line

File: src/Startup.cs [hunk 8]
```diff
@@ -33,7 +36,14 @@ public void ConfigureServices(IServiceCollection services)
 
 
             services.AddMvc().SetCompatibilityVersion(CompatibilityVersion.Version_2_1);
-        }
+            services.AddDbContext<DataBaseContext>(options =>  options.UseNpgsql(Configuration["ConnectionString"]));
```
Comment (gzamudio) [inline src/Startup.cs]: remove extra space before 'options.Use..'

File: src/Startup.cs [hunk 9]
```diff
@@ -33,7 +36,14 @@ public void ConfigureServices(IServiceCollection services)
...
 
             services.AddMvc().SetCompatibilityVersion(CompatibilityVersion.Version_2_1);
-        }
+            services.AddDbContext<DataBaseContext>(options =>  options.UseNpgsql(Configuration["ConnectionString"]));
+            services.AddScoped<DataBaseContext>();
+
+            services.AddJsonLocalization(options => options.ResourcesPath = "Resources");
+            services.AddMvc().AddViewLocalization();
+            CultureInfo.CurrentUICulture = new CultureInfo(Configuration["DefaultLang"]);   
+
```
Comment (gzamudio) [inline src/Startup.cs]: remove empty line

File: src/Startup.cs [hunk 10]
```diff
@@ -33,7 +36,14 @@ public void ConfigureServices(IServiceCollection services)
...
             services.AddMvc().SetCompatibilityVersion(CompatibilityVersion.Version_2_1);
-        }
+            services.AddDbContext<DataBaseContext>(options =>  options.UseNpgsql(Configuration["ConnectionString"]));
+            services.AddScoped<DataBaseContext>();
+
+            services.AddJsonLocalization(options => options.ResourcesPath = "Resources");
+            services.AddMvc().AddViewLocalization();
+            CultureInfo.CurrentUICulture = new CultureInfo(Configuration["DefaultLang"]);   
+            
+            }
```
Comment (gzamudio) [inline src/Startup.cs]: check indentation

File: src/Views/HelloWorld/Welcome.cshtml [hunk 11]
```diff
@@ -0,0 +1,12 @@
...
+}
+
+<h2>Welcome</h2>
+
+<ul>
+    @for (int i = 0; i < (int)ViewData["NumTimes"]; i++)
+    {
+        <li>@ViewData["Message"]</li>
+    }
+</ul>
```
Comment (gzamudio) [inline src/Views/HelloWorld/Welcome.cshtml]: add empty line

File: src/testing-net.csproj [hunk 12]
```diff
@@ -1,11 +1,10 @@
...
+<Project Sdk="Microsoft.NET.Sdk.Web">
+  <PropertyGroup>
+    <TargetFramework>netcoreapp2.1</TargetFramework>
+  </PropertyGroup>
+  <ItemGroup>
+    <PackageReference Include="Microsoft.AspNetCore.App"/>
+    <PackageReference Include="Npgsql.EntityFrameworkCore.PostgreSQL" Version="2.1.1.1"/>
+    <PackageReference Include="LocalizationCultureCore" Version="1.1.2"/>
+  </ItemGroup>
+</Project>
```
Comment (gzamudio) [inline src/testing-net.csproj]: add an empty line

**Review by gzamudio — CHANGES_REQUESTED** [review]

File: src/testing-net.csproj (hunk 12 above)
Comment (j1nma) [reply src/testing-net.csproj]: Done.

File: src/Views/HelloWorld/Welcome.cshtml (hunk 11 above)
Comment (j1nma) [reply src/Views/HelloWorld/Welcome.cshtml]: Done.

File: src/Resources/en-US.json (hunk 5 above)
Comment (j1nma) [reply src/Resources/en-US.json]: Done.

File: src/Resources/es-MX.json (hunk 6 above)
Comment (j1nma) [reply src/Resources/es-MX.json]: Done.

File: src/Resources/fr-FR.json (hunk 7 above)
Comment (j1nma) [reply src/Resources/fr-FR.json]: Done.

File: src/Controllers/MoviesController.cs [hunk 13]
```diff
@@ -0,0 +1,13 @@
+using System.Text.Encodings.Web;
+using Microsoft.AspNetCore.Mvc;
+
+namespace testing_net.Controllers 
+{
+    public class MoviesController : Controller 
+    {
+        public IActionResult Index () 
+        {
+            return View ();
```
Comment (gzamudio) [inline src/Controllers/MoviesController.cs]: remove the space before the parenthesis

File: src/Controllers/HomeController.cs [hunk 14]
```diff
@@ -1,43 +1,52 @@
...
+        public IActionResult Contact () 
+        {
+            ViewData["Message"] = _localizer["ContactPage"];
+
+            return View ();
+        }
+
+        public IActionResult Privacy () 
+        {
+            return View ();
```
Comment (gzamudio) [inline src/Controllers/HomeController.cs]: remove the space before the parenthesis

File: src/Controllers/HomeController.cs [hunk 15]
```diff
@@ -1,43 +1,52 @@
...
+        }
+
+        [ResponseCache (Duration = 0, Location = ResponseCacheLocation.None, NoStore = true)]
+        public IActionResult Error () 
+        {
+            return View (new ErrorViewModel { RequestId = Activity.Current?.Id ?? HttpContext.TraceIdentifier });
```
Comment (gzamudio) [inline src/Controllers/HomeController.cs]: remove the space before the parenthesis

File: src/Controllers/HomeController.cs [hunk 16]
```diff
@@ -1,43 +1,52 @@
...
+        {
+            ViewData["Message"] = _localizer["DescriptionPage"];
+
+            return View ();
+        }
+
+        public IActionResult Contact () 
+        {
+            ViewData["Message"] = _localizer["ContactPage"];
+
```
Comment (gzamudio) [inline src/Controllers/HomeController.cs]: remove empty line

File: src/Controllers/HomeController.cs [hunk 17]
```diff
@@ -1,43 +1,52 @@
...
+        public IActionResult Index () 
+        {
+
+            return View ();
+        }
+
+        public IActionResult About () 
+        {
+            ViewData["Message"] = _localizer["DescriptionPage"];
+
```
Comment (gzamudio) [inline src/Controllers/HomeController.cs]: remove empty line

File: src/Controllers/HomeController.cs [hunk 18]
```diff
@@ -1,43 +1,52 @@
...
+        private readonly IHtmlLocalizer<HomeController> _localizer;
+
+        public HomeController (IHtmlLocalizer<HomeController> localizer) 
+        {
+            this._localizer = localizer;
+        }
+
+        public IActionResult Index () 
+        {
+
```
Comment (gzamudio) [inline src/Controllers/HomeController.cs]: remove empty line

File: src/Controllers/HelloWorldController.cs [hunk 19]
```diff
@@ -1,23 +1,21 @@
...
 
-        // GET: /HelloWorld/Welcome/ 
-        // Requires using System.Text.Encodings.Web;
-        public string Welcome(string name, int ID = 1)
+        public IActionResult Welcome (string name, int numTimes = 1) 
         {
-            return HtmlEncoder.Default.Encode($"Hello {name}, ID: {ID}");
+            ViewData["Message"] = "Hello " + name;
+            ViewData["NumTimes"] = numTimes;
+
```
Comment (gzamudio) [inline src/Controllers/HelloWorldController.cs]: remove empty line

File: src/Controllers/HelloWorldController.cs [hunk 20]
```diff
@@ -1,23 +1,21 @@
...
     public class HelloWorldController : Controller
     {
-        // 
-        // GET: /HelloWorld/
-
-        public string Index()
+        public IActionResult Index () 
         {
-            return "This is my default action...";
+            return View ();
```
Comment (gzamudio) [inline src/Controllers/HelloWorldController.cs]: remove the space before the parenthesis

**Review by gzamudio — APPROVED** [review]
there are a few comments. fix the spaces and empty lines and merge. we only use empty lines at the end of files, and between methods or properties.

**Pull request merged**